bpf/lib/process.h

```diff
@@ -535,4 +535,42 @@ execve_joined_info_map_get(__u64 tid)
 
 _Static_assert(sizeof(struct execve_map_value) % 8 == 0,
 	       "struct execve_map_value should have size multiple of 8 bytes");
+
+#define METRIC_EXEC	  0
+#define METRIC_CLONE	  1
+#define METRIC_EXIT	  2
+#define METRIC_DATA	  3
+#define METRIC_CGROUP	  4
+#define METRIC_LOADER	  5
+#define METRIC_TRACEPOINT 6
+#define METRIC_KPROBE	  7
+#define METRIC_UPROBE	  8
+#define METRIC_MAX_VALUES 9
+
+struct kernel_stats {
+	__u64 sent[METRIC_MAX_VALUES];
+	__u64 sent_failed[METRIC_MAX_VALUES];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, __u32);
+	__type(value, struct kernel_stats);
+	__uint(max_entries, 1);
+} tg_stats_map SEC(".maps");
+
+#define PERF_METRICS(__x, __v)                                                      \
+	do {                                                                        \
+		long retval = (__v);                                                \
+		long index = (__x);                                                 \
+		__u32 key = 0;                                                      \
+		struct kernel_stats *valp = map_lookup_elem(&tg_stats_map, &key);   \
+		if (valp) {                                                         \
+			if (retval >= 0)                                            \
+				__sync_fetch_and_add(&valp->sent[index], 1);        \
```

Comment:
Hm... do we need this? It adds overhead in the good path and we count the events we receive by type in user-space.

Reply:
I have seen cases where the sum of all `tetragon_missed_events_total` is not equal to `tetragon_ringbuf_perf_event_lost_total` under very heavy load. That meant to me that there were some events that we successfully sent from the kernel and somehow we missed getting into the user space. 

I have tried to reproduce that now with the latest version but I didn't manage to do so. So let's remove the `tetragon_sent_events_total` metric entirely for now and we can reconsider that in the future if we see again similar cases. 

Regarding the overhead, I don't believe that this is a big deal as we use a per-cpu array.

Reply:
> I have seen cases where the sum of all tetragon_missed_events_total is not equal to tetragon_ringbuf_perf_event_lost_total under very heavy load. That meant to me that there were some events that we successfully sent from the kernel and somehow we missed getting into the user space

OK, that's good enough reason to keep it for me.
If we use a per-cpu array, do we need `__sync_fetch_and_add`?

Reply:
> OK, that's good enough reason to keep it for me.

I will try to do that again. If I don't manage to do that I will remove that for now (I am just thinking if there was any bugs in the previous version and thus I have this behaviour). If I don't manage to reproduce, I will remove that for now. 

> If we use a per-cpu array, do we need __sync_fetch_and_add?

Hm, I would say no but let's be in the safe side. I will try to experiment with that as well a bit.

Reply:
After some experiments:
1. Let's remove the metric for successful `perf_event_output` calls. Maybe I have seen a bug in a previous version. Even with a very loaded host, I cannot reproduce that. We can always add that again if we see any numbers that do not make sense.
2. I don't really believe that `__sync_fetch_and_add` is necessary, but I would say let's keep it. It is called only when `perf_event_output` fails which is a rare event. Thus any performance overhead is minimal here. 

Does this make sense?

Reply:
Sounds good to me, thanks!